File: base-v3/src/main/java/com/adyen/checkout/base/util/PaymentMethodTypes.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 4/3/2019.
 */

package com.adyen.checkout.base.util;

import com.adyen.checkout.core.exception.NoConstructorException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Helper class with a list of all the currently supported Payment Methods on Components and Drop-In.
 */
@SuppressWarnings("SpellCheckingInspection")
public final class PaymentMethodTypes {

    // Placeholder value if the type is not found.
    public static final String UNKNOWN = "unknown";

    // Type of the payment method as received by the paymentMethods/ API
    public static final String IDEAL = "ideal";
    public static final String MOLPAY_MALAYSIA = "molpay_ebanking_fpx_MY";
    public static final String MOLPAY_THAILAND = "molpay_ebanking_TH";
    public static final String MOLPAY_VIETNAM = "molpay_ebanking_VN";
    public static final String DOTPAY = "dotpay";
    public static final String EPS = "eps";
    public static final String ENTERCASH = "entercash";
    public static final String OPEN_BANKING = "openbanking_UK";
    public static final String SCHEME = "scheme";
    public static final String GOOGLE_PAY = "paywithgoogle";
    public static final String SEPA = "sepadirectdebit";
    public static final String AFTER_PAY = "afterpay_default";
    public static final String BCMC = "bcmc";
    public static final String WECHAT_PAY_SDK = "wechatpaySDK";
    public static final String MB_WAY = "mbway";

    // Voucher payment methods that are not yet supported
    public static final String MULTIBANCO = "multibanco";
    public static final String OXXO = "oxxo";

    public static final String DOKU = "doku";
    public static final String DOKU_ALFMART = "doku_alfamart";
    public static final String DOKU_PERMATA_LITE_ATM = "doku_permata_lite_atm";
    public static final String DOKU_INDOMARET = "doku_indomaret";
    public static final String DOKU_ATM_MANDIRI_VA = "doku_atm_mandiri_va";
    public static final String DOKU_SINARMAS_VA = "doku_sinarmas_va";
    public static final String DOKU_MANDIRI_VA = "doku_mandiri_va";
    public static final String DOKU_CIMB_VA = "doku_cimb_va";
    public static final String DOKU_DANAMON_VA = "doku_danamon_va";
    public static final String DOKU_BRI_VA = "doku_bri_va";
    public static final String DOKU_BNI_VA = "doku_bni_va";
    public static final String DOKU_BCA_VA = "doku_bca_va";
    public static final String DOKU_WALLET = "doku_wallet";

    public static final String BOLETOBANCARIO = "boletobancario";
    public static final String BOLETOBANCARIO_BANCODOBRASIL = "boletobancario_bancodobrasil";
    public static final String BOLETOBANCARIO_BRADESCO = "boletobancario_bradesco";
    public static final String BOLETOBANCARIO_HSBC = "boletobancario_hsbc";
    public static final String BOLETOBANCARIO_ITAU = "boletobancario_itau";
    public static final String BOLETOBANCARIO_SANTANDER = "boletobancario_santander";

    public static final String DRAGONPAY_EBANKING = "dragonpay_ebanking";
    public static final String DRAGONPAY_OTC_BANKING = "dragonpay_otc_banking";
    public static final String DRAGONPAY_OTC_NON_BANKING = "dragonpay_otc_non_banking";
    public static final String DRAGONPAY_OTC_PHILIPPINES = "dragonpay_otc_philippines";

    public static final String ECONTEXT_SEVEN_ELEVEN = "econtext_seven_eleven";
    public static final String ECONTEXT_ATM = "econtext_atm";
    public static final String ECONTEXT_STORES = "econtext_stores";
    public static final String ECONTEXT_ONLINE = "econtext_online";

    // Payment methods that might be interpreted as redirect, but are actually not supported
    public static final String BCMC_QR = "bcmc_mobile_QR";
    public static final String WECHAT_PAY_MINI_PROGRAM = "wechatpayMiniProgram";
    public static final String WECHAT_PAY_QR = "wechatpayQR";
    public static final String WECHAT_PAY_WEB = "wechatpayWeb";

    // List of all payment method types.
    public static final List<String> SUPPORTED_PAYMENT_METHODS;
    public static final List<String> UNSUPPORTED_PAYMENT_METHODS;

    static {
        final ArrayList<String> supportedPaymentMethods = new ArrayList<>();

        // Populate supported list
        supportedPaymentMethods.add(AFTER_PAY);
        supportedPaymentMethods.add(BCMC);
        supportedPaymentMethods.add(DOTPAY);
        supportedPaymentMethods.add(ENTERCASH);
        supportedPaymentMethods.add(EPS);
        supportedPaymentMethods.add(GOOGLE_PAY);
        supportedPaymentMethods.add(IDEAL);
        supportedPaymentMethods.add(MB_WAY);
        supportedPaymentMethods.add(MOLPAY_MALAYSIA);
        supportedPaymentMethods.add(MOLPAY_THAILAND);
        supportedPaymentMethods.add(MOLPAY_VIETNAM);
        supportedPaymentMethods.add(OPEN_BANKING);
        // Sepa is not supported until we fix the Component
        //    supportedPaymentMethods.add(SEPA);
        supportedPaymentMethods.add(SCHEME);
        supportedPaymentMethods.add(WECHAT_PAY_SDK);

        SUPPORTED_PAYMENT_METHODS = Collections.unmodifiableList(supportedPaymentMethods);

        final ArrayList<String> unsupportedPaymentMethods = new ArrayList<>();

        // Populate unsupported list

        // Sepa is not supported until we fix the Component
        unsupportedPaymentMethods.add(SEPA);

        unsupportedPaymentMethods.add(BCMC_QR);
        unsupportedPaymentMethods.add(WECHAT_PAY_MINI_PROGRAM);
        unsupportedPaymentMethods.add(WECHAT_PAY_QR);
        unsupportedPaymentMethods.add(WECHAT_PAY_WEB);

        unsupportedPaymentMethods.add(MULTIBANCO);
        unsupportedPaymentMethods.add(OXXO);
        unsupportedPaymentMethods.add(DOKU);
        unsupportedPaymentMethods.add(DOKU_ALFMART);
        unsupportedPaymentMethods.add(DOKU_PERMATA_LITE_ATM);
        unsupportedPaymentMethods.add(DOKU_INDOMARET);
        unsupportedPaymentMethods.add(DOKU_ATM_MANDIRI_VA);
        unsupportedPaymentMethods.add(DOKU_SINARMAS_VA);
        unsupportedPaymentMethods.add(DOKU_MANDIRI_VA);
        unsupportedPaymentMethods.add(DOKU_CIMB_VA);
        unsupportedPaymentMethods.add(DOKU_DANAMON_VA);
        unsupportedPaymentMethods.add(DOKU_BRI_VA);
        unsupportedPaymentMethods.add(DOKU_BNI_VA);
        unsupportedPaymentMethods.add(DOKU_BCA_VA);
        unsupportedPaymentMethods.add(DOKU_WALLET);

        unsupportedPaymentMethods.add(BOLETOBANCARIO);
        unsupportedPaymentMethods.add(BOLETOBANCARIO_BANCODOBRASIL);
        unsupportedPaymentMethods.add(BOLETOBANCARIO_BRADESCO);
        unsupportedPaymentMethods.add(BOLETOBANCARIO_HSBC);
        unsupportedPaymentMethods.add(BOLETOBANCARIO_ITAU);
        unsupportedPaymentMethods.add(BOLETOBANCARIO_SANTANDER);

        unsupportedPaymentMethods.add(DRAGONPAY_EBANKING);
        unsupportedPaymentMethods.add(DRAGONPAY_OTC_BANKING);
        unsupportedPaymentMethods.add(DRAGONPAY_OTC_NON_BANKING);
        unsupportedPaymentMethods.add(DRAGONPAY_OTC_PHILIPPINES);

        unsupportedPaymentMethods.add(ECONTEXT_SEVEN_ELEVEN);
        unsupportedPaymentMethods.add(ECONTEXT_ATM);
        unsupportedPaymentMethods.add(ECONTEXT_STORES);
        unsupportedPaymentMethods.add(ECONTEXT_ONLINE);

        UNSUPPORTED_PAYMENT_METHODS = Collections.unmodifiableList(unsupportedPaymentMethods);
    }

    private PaymentMethodTypes() {
        throw new NoConstructorException();
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/core/PaymentMethodTypes.kt
/*
 * Copyright (c) 2023 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 13/3/2023.
 */
package com.adyen.checkout.components.core

/**
 * Helper class with a list of all the currently supported Payment Methods on Components and Drop-In.
 */
object PaymentMethodTypes {

    // Placeholder value if the type is not found.
    const val UNKNOWN = "unknown"

    // Type of the payment method as received by the paymentMethods/ API
    const val ACH = "ach"
    const val BACS = "directdebit_GB"
    const val BCMC = "bcmc"
    const val BLIK = "blik"
    const val BOLETOBANCARIO = "boletobancario"
    const val BOLETOBANCARIO_BANCODOBRASIL = "boletobancario_bancodobrasil"
    const val BOLETOBANCARIO_BRADESCO = "boletobancario_bradesco"
    const val BOLETOBANCARIO_HSBC = "boletobancario_hsbc"
    const val BOLETOBANCARIO_ITAU = "boletobancario_itau"
    const val BOLETOBANCARIO_SANTANDER = "boletobancario_santander"
    const val BOLETO_PRIMEIRO_PAY = "primeiropay_boleto"
    const val CASH_APP_PAY = "cashapp"
    const val DOTPAY = "dotpay"
    const val ENTERCASH = "entercash"
    const val EPS = "eps"
    const val GIFTCARD = "giftcard"
    const val GOOGLE_PAY = "googlepay"
    const val GOOGLE_PAY_LEGACY = "paywithgoogle"
    const val IDEAL = "ideal"
    const val MB_WAY = "mbway"
    const val MEAL_VOUCHER_FR_GROUPEUP = "mealVoucher_FR_groupeup"
    const val MEAL_VOUCHER_FR_NATIXIS = "mealVoucher_FR_natixis"
    const val MEAL_VOUCHER_FR_SODEXO = "mealVoucher_FR_sodexo"
    const val MEAL_VOUCHER_FR = "mealVoucher_FR"
    const val MOLPAY_MALAYSIA = "molpay_ebanking_fpx_MY"
    const val MOLPAY_THAILAND = "molpay_ebanking_TH"
    const val MOLPAY_VIETNAM = "molpay_ebanking_VN"
    const val ONLINE_BANKING_CZ = "onlineBanking_CZ"
    const val ONLINE_BANKING_PL = "onlineBanking_PL"
    const val ONLINE_BANKING_SK = "onlineBanking_SK"
    const val OPEN_BANKING = "openbanking_UK"
    const val PAY_BY_BANK = "paybybank"
    const val PAY_BY_BANK_US = "paybybank_AIS_DD"
    const val SCHEME = "scheme"
    const val SEPA = "sepadirectdebit"
    const val TWINT = "twint"
    const val UPI = "upi"
    const val UPI_INTENT = "upi_intent"
    const val UPI_COLLECT = "upi_collect"
    const val UPI_QR = "upi_qr"

    // Payment methods that do not need a payment component, but only an action component
    const val DUIT_NOW = "duitnow"
    const val PAY_NOW = "paynow"
    const val PIX = "pix"
    const val PROMPT_PAY = "promptpay"
    const val WECHAT_PAY_SDK = "wechatpaySDK"
    const val MULTIBANCO = "multibanco"

    // Voucher payment methods that are not yet supported
    const val DOKU = "doku"
    const val DOKU_ALFMART = "doku_alfamart"
    const val DOKU_ATM_MANDIRI_VA = "doku_atm_mandiri_va"
    const val DOKU_BCA_VA = "doku_bca_va"
    const val DOKU_BNI_VA = "doku_bni_va"
    const val DOKU_BRI_VA = "doku_bri_va"
    const val DOKU_CIMB_VA = "doku_cimb_va"
    const val DOKU_DANAMON_VA = "doku_danamon_va"
    const val DOKU_INDOMARET = "doku_indomaret"
    const val DOKU_MANDIRI_VA = "doku_mandiri_va"
    const val DOKU_PERMATA_LITE_ATM = "doku_permata_lite_atm"
    const val DOKU_SINARMAS_VA = "doku_sinarmas_va"
    const val DOKU_WALLET = "doku_wallet"
    const val DRAGONPAY_EBANKING = "dragonpay_ebanking"
    const val DRAGONPAY_OTC_BANKING = "dragonpay_otc_banking"
    const val DRAGONPAY_OTC_NON_BANKING = "dragonpay_otc_non_banking"
    const val DRAGONPAY_OTC_PHILIPPINES = "dragonpay_otc_philippines"
    const val ECONTEXT_ATM = "econtext_atm"
    const val ECONTEXT_ONLINE = "econtext_online"
    const val ECONTEXT_SEVEN_ELEVEN = "econtext_seven_eleven"
    const val ECONTEXT_STORES = "econtext_stores"
    const val OXXO = "oxxo"

    // Payment methods that might be interpreted as redirect, but are actually not supported
    const val AFTER_PAY = "afterpay_default"
    const val BCMC_QR = "bcmc_mobile_QR"
    const val WECHAT_PAY_MINI_PROGRAM = "wechatpayMiniProgram"
    const val WECHAT_PAY_QR = "wechatpayQR"
    const val WECHAT_PAY_WEB = "wechatpayWeb"

    // List of all payment method types.
    val SUPPORTED_PAYMENT_METHODS: List<String> = listOf(
        ACH,
        BACS,
        BCMC,
        BLIK,
        BOLETOBANCARIO,
        BOLETOBANCARIO_BANCODOBRASIL,
        BOLETOBANCARIO_BRADESCO,
        BOLETOBANCARIO_HSBC,
        BOLETOBANCARIO_ITAU,
        BOLETOBANCARIO_SANTANDER,
        BOLETO_PRIMEIRO_PAY,
        CASH_APP_PAY,
        DOTPAY,
        DUIT_NOW,
        ECONTEXT_ATM,
        ECONTEXT_ONLINE,
        ECONTEXT_SEVEN_ELEVEN,
        ECONTEXT_STORES,
        ENTERCASH,
        EPS,
        GIFTCARD,
        GOOGLE_PAY,
        GOOGLE_PAY_LEGACY,
        IDEAL,
        MB_WAY,
        MEAL_VOUCHER_FR_GROUPEUP,
        MEAL_VOUCHER_FR_NATIXIS,
        MEAL_VOUCHER_FR_SODEXO,
        MOLPAY_MALAYSIA,
        MOLPAY_THAILAND,
        MOLPAY_VIETNAM,
        MULTIBANCO,
        ONLINE_BANKING_CZ,
        ONLINE_BANKING_PL,
        ONLINE_BANKING_SK,
        OPEN_BANKING,
        PAY_BY_BANK,
        PAY_BY_BANK_US,
        PAY_NOW,
        PIX,
        PROMPT_PAY,
        SCHEME,
        SEPA,
        TWINT,
        UPI,
        UPI_COLLECT,
        UPI_INTENT,
        UPI_QR,
        WECHAT_PAY_SDK,
    )

    // Payment methods that do not need a payment component, but only an action component
    @Suppress("unused")
    @Deprecated(
        """
        This list is no longer relevant nor maintained. To check which payment methods are supported by the 
        InstantPayComponent use InstantPayComponent.PROVIDER.isPaymentMethodSupported.""",
    )
    val SUPPORTED_ACTION_ONLY_PAYMENT_METHODS: List<String> = listOf(
        DUIT_NOW,
        PAY_NOW,
        PIX,
        PROMPT_PAY,
        TWINT,
        WECHAT_PAY_SDK,
        MULTIBANCO,
    )

    // Payment methods that are explicitly unsupported
    val UNSUPPORTED_PAYMENT_METHODS: List<String> = listOf(
        AFTER_PAY,
        BCMC_QR,
        DOKU,
        DOKU_ALFMART,
        DOKU_ATM_MANDIRI_VA,
        DOKU_BCA_VA,
        DOKU_BNI_VA,
        DOKU_BRI_VA,
        DOKU_CIMB_VA,
        DOKU_DANAMON_VA,
        DOKU_INDOMARET,
        DOKU_MANDIRI_VA,
        DOKU_PERMATA_LITE_ATM,
        DOKU_SINARMAS_VA,
        DOKU_WALLET,
        DRAGONPAY_EBANKING,
        DRAGONPAY_OTC_BANKING,
        DRAGONPAY_OTC_NON_BANKING,
        DRAGONPAY_OTC_PHILIPPINES,
        OXXO,
        WECHAT_PAY_MINI_PROGRAM,
        WECHAT_PAY_QR,
        WECHAT_PAY_WEB,
    )
}
